cylc/uiserver/authorise.py

```diff
@@ -0,0 +1,477 @@
+                        for conf in [existing_limit, new_limit]:
+                            if isinstance(conf, list):
+                                set_lims.update(conf)
+                            else:
+                                set_lims.add(conf)
+                        # update and continue
+                        owner_dict[
+                            acc_user_conf][Authorization.LIMIT] = list(
+                            set_lims)
+                        owner_dict[
+                            acc_user_conf][Authorization.DEFAULT] = list(
+                            set_defs)
+                        continue
+                    owner_dict.update(access_user_dict)
+        # Now we have a reduced site auth dictionary for the current owner
+        return owner_dict
+
+    def return_site_auth_defaults_for_access_user(
+            self, access_user: Dict[str, Union[str, Sequence[Any]]]) -> Set:
+        """Return site authorization defaults for given access user.
+            Args:
+                access_user: access_user dictionary, in the form
+                            {'access_username': username
+                             'access_user_group: [group1, group2,...]'
+                            }
+            Returns:
+                Set of default operations permitted
+
+        """
+        defaults: Set[str] = set()
+        if not self.owner_dict:
+            return defaults
+        items_to_check = ['*', access_user['access_username']]
+        items_to_check.extend(access_user['access_user_groups'])
+        for item in items_to_check:
+            permission = ''
+            with suppress(KeyError):
+                permission = self.owner_dict[item].get(
+                    Authorization.DEFAULT, ''
+                )
+            if isinstance(permission, str):
+                defaults.add(permission)
+            else:
+                defaults.update(permission)
+        defaults.discard('')
+        return defaults
+
+
+# GraphQL middleware
+class AuthorizationMiddleware:
+
+    auth = None
+    current_user = None
+
+    def resolve(self, next_, root, info, **args):
+        # We won't be re-checking auth for return variables
+        if len(info.path) > 1:
+            return next_(root, info, **args)
+        op_name = self.get_op_name(info.field_name, info.operation.operation)
+        # It shouldn't get here but worth checking for zero trust
+        if not op_name:
+            self.auth_failed(self.current_user, op_name, 400,
+                             "Operation not in schema.")
+        authorised = self.auth.is_permitted(self.current_user, op_name)
+        if not authorised:
+            self.auth_failed(self.current_user, op_name, 403)
+        if (info.operation.operation in Authorization.ASYNC_OPS
+                or iscoroutinefunction(next_)):
+            return self.async_resolve(next_, root, info, **args)
+        return next_(root, info, **args)
```

Comment:
I will need :eyeglasses: + :coffee: + IDE/debugger to review this function next week